src/CharterManager.sol

```diff
@@ -0,0 +1,256 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+// Copyright (C) 2021 Dai Foundation
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+pragma solidity 0.6.12;
+
+interface VatLike {
+    function live() external view returns (uint256);
+    function urns(bytes32, address) external view returns (uint256, uint256);
+    function dai(address) external view returns (uint256);
+    function fork(bytes32, address, address, int256, int256) external;
+    function frob(bytes32, address, address, address, int256, int256) external;
+    function flux(bytes32, address, address, uint256) external;
+    function move(address, address, uint256) external;
+    function hope(address) external;
+    function nope(address) external;
+    function ilks(bytes32) external view returns (
+        uint256 Art,  // [wad]
+        uint256 rate, // [ray]
+        uint256 spot, // [ray]
+        uint256 line, // [rad]
+        uint256 dust  // [rad]
+    );
+}
+
+interface ManagedGemJoinLike {
+    function gem() external view returns (address);
+    function ilk() external view returns (bytes32);
+    function join(address, uint256) external;
+    function exit(address, address, uint256) external;
+}
+
+interface TokenLike {
+    function approve(address, uint256) external;
+    function transferFrom(address, address, uint256) external;
+}
+
+contract UrnProxy {
+    address immutable public usr;
+
+    constructor(address vat_, address usr_) public {
+        usr = usr_;
+        VatLike(vat_).hope(msg.sender);
+    }
+}
+
+contract CharterManager {
+    mapping (address => uint256) public wards;
+    address public implementation;
```

Comment:
This is minor and subjective thing I changed in CropJoin, basically moving the `implementation` to slot0 as this should never be used in the `implementation` itself. So that way it doesn't get in the middle of variables that will be used.
Anyway whatever is preferred. On the other hand in the regular contracts we use to have `wards` always in slot 0.

Reply:
Yeah, I like this change.
I actually wasn't sure it would work as I thought we might rely on `wards` being in the first slot.
But was pleasently surprised that the testing utility `giveAuth` searches for the `wards` location.
Done.